GPUP: Gradient identifiers should be handled in type-safe manner

kkinnunen-apple opened the pull request:
#### fa0ef97b339c27752199f878df1dd1088e1c220a
<pre>
GPUP: Gradient identifiers should be handled in type-safe manner
<a href="https://bugs.webkit.org/show_bug.cgi?id=298650">https://bugs.webkit.org/show_bug.cgi?id=298650</a>
<a href="https://rdar.apple.com/160274124">rdar://160274124</a>

Reviewed by Said Abou-Hallawa.

Introduce WebKit::RemoteGradientIdentifier to identify Gradients, instead of
the overloaded WebCore::RemoteResourceIdentifier. Remote aspects
should be in WebKit namespace.

* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::create):
(WebCore::Gradient::Gradient):
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/Gradient.h:
(isType): Deleted.
* Source/WebCore/platform/graphics/RenderingResource.h:
(WebCore::RenderingResource::isNativeImage const):
(WebCore::RenderingResource::isGradient const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp:
(WebCore::RenderSVGResourceLinearGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.cpp:
(WebCore::RenderSVGResourceRadialGradient::createGradient):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceLinearGradient.cpp:
(WebCore::LegacyRenderSVGResourceLinearGradient::buildGradient const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceRadialGradient.cpp:
(WebCore::LegacyRenderSVGResourceRadialGradient::buildGradient const):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.cpp:
(WebKit::RemoteGraphicsContext::setFillCachedGradient):
(WebKit::RemoteGraphicsContext::setStrokeCachedGradient):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheGradient):
(WebKit::RemoteRenderingBackend::releaseGradient):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheGradient):
(WebKit::RemoteResourceCache::releaseGradient):
(WebKit::RemoteResourceCache::cachedGradient const):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
(WebKit::RemoteGraphicsContextProxy::appendStateChangeItemIfNecessary):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheGradient):
(WebKit::RemoteRenderingBackendProxy::releaseGradient):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordGradientUse):
(WebKit::RemoteResourceCacheProxy::willDestroyGradient):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: <a href="https://commits.webkit.org/300094@main">https://commits.webkit.org/300094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba5ad4dba50580eea751594e8ccf6946078a3fd2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127740 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73385 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/59d94eb3-057a-4967-8a57-98f771490692) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123180 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92150 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61306 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/34432b4c-0beb-422e-8ce5-c8b01f54113c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124256 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33297 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108705 "Found 1 new API test failure: TestIPC.StreamServerConnectionTests/StreamServerDidReceiveInvalidMessageTest.AsyncNotStreamEncodable/DecodeError (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72826 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cf6fd981-6247-4ca0-93da-c521d9ee2c64) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/120661 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32314 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26833 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71323 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130578 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100746 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100651 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25511 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46054 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24117 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44926 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48090 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53803 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47562 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50908 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49244 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/Gradient.h:
```diff
@@ -132,12 +138,10 @@ class Gradient final : public RenderingResource {
 #endif
 
+    mutable WeakHashSet<RenderingResourceObserver> m_observers;
+    bool m_isTransient { true };
```
Can you add a comment explaining what "isTransient" means here?

kkinnunen-apple replied:
Thanks, added..

shallawa commented on Source/WebCore/platform/graphics/Gradient.cpp:
```diff
@@ -35,24 +35,24 @@
 namespace WebCore {
 
-Ref<Gradient> Gradient::create(Data&& data, ColorInterpolationMethod colorInterpolationMethod, GradientSpreadMethod spreadMethod, GradientColorStops&& stops, std::optional<RenderingResourceIdentifier> renderingResourceIdentifier)
+Ref<Gradient> Gradient::create(Data&& data, ColorInterpolationMethod colorInterpolationMethod, GradientSpreadMethod spreadMethod, GradientColorStops&& stops, bool isTransient)
```
Instead of using `bool isTransient`, cannot we have a separate function for creating transient gradient e.g. `Gradient::createTransient()`?

kkinnunen-apple replied:
Done.

kkinnunen-apple replied:
In fact, did not add this one. The ipc decode system cannot use such a constructor, so its maybe best to have a simple one constructor

shallawa commented on Source/WebCore/platform/graphics/Gradient.h:
```diff
@@ -83,13 +85,17 @@ class Gradient final : public RenderingResource {
 
-    WEBCORE_EXPORT static Ref<Gradient> create(Data&&, ColorInterpolationMethod, GradientSpreadMethod = GradientSpreadMethod::Pad, GradientColorStops&& = { }, std::optional<RenderingResourceIdentifier> = std::nullopt);
-    ~Gradient();
+    // isTransient may affect backend rendering implementation caching decisions.
+    // Transient instances may be assumed to be drawn only few times or seldomly and as such the backend
+    // may not persist caches related to the instance.
```
A transient gradient does not need to be cached. It is usually created before filling or stroking and deleted after the drawing.

shallawa commented on Source/WebCore/platform/graphics/Gradient.h:
```diff
@@ -115,10 +121,13 @@ class Gradient final : public RenderingResource {
 
-private:
-    Gradient(Data&&, ColorInterpolationMethod, GradientSpreadMethod, GradientColorStops&&, std::optional<RenderingResourceIdentifier>);
+    void addObserver(WeakRef<RenderingResourceObserver>&& observer) const
+    {
+        m_observers.add(WTFMove(observer));
+    }
```
What is the plan of the `RenderingResourceObservers`? Is the plan to add similar code in all types of resources. I think this code can be put in some base class instead of repeating it.

kkinnunen-apple replied:
The idea is to not have a base class, because the types themselves are not related. They also might have slightly different refcounting or weakptr behavior. The code itself is trivial, I do not think there is anything gained by having a the simple code in a shared mixin. 

Later on, the idea is perhaps to replace RenderingResourceObserver with more fine-grained observers, notably to NativeImageObserver  and perhaps RenderingResourceObserver. This is unclear ATM, since it's not trivially clear to me if the multiple weakptr interfaces work then with the same receiver. 

The NativeImage counting needs are different to others currently, so it may be that fine-grained observer would be useful.

If the fine-grained observer is not needed, the mixing approach (not a shared base class) could be reintroduced.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
```diff
@@ -82,7 +83,7 @@ class RemoteResourceCacheProxy final : public WebCore::RenderingResourceObserver
 
     HashSet<WebCore::RenderingResourceIdentifier> m_nativeImages;
-    HashSet<WebCore::RenderingResourceIdentifier> m_gradients;
+    HashMap<const WebCore::Gradient*, RemoteGradientIdentifier> m_gradients;
```
Why is this a raw pointer? Should not it be a `WeakPtr` or `WeakRef`?

kkinnunen-apple replied:
The pointer is used to identify the object. We can do this, because we subscribe to destruction and remove the instance from the map. The pointer itself is not dereferenced. We could use OpaquePtrValue<const WebCore::Gradient> (a strongly typed intptr_t), but we don't have that type in WTF at the moment. I could add that later if needed..

Gradient is not weakptr type, the feature along its overhead is not needed. Since it's not needed, not adding it is better for less mental overhead.

kkinnunen-apple committed:
GPUP: Gradient identifiers should be handled in type-safe manner
https://bugs.webkit.org/show_bug.cgi?id=298650
rdar://160274124

Reviewed by Said Abou-Hallawa.

Introduce WebKit::RemoteGradientIdentifier to identify Gradients, instead of
the overloaded WebCore::RemoteResourceIdentifier. Remote aspects
should be in WebKit namespace.

* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::create):
(WebCore::Gradient::Gradient):
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/Gradient.h:
(isType): Deleted.
* Source/WebCore/platform/graphics/RenderingResource.h:
(WebCore::RenderingResource::isNativeImage const):
(WebCore::RenderingResource::isGradient const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceLinearGradient.cpp:
(WebCore::RenderSVGResourceLinearGradient::createGradient):
* Source/WebCore/rendering/svg/RenderSVGResourceRadialGradient.cpp:
(WebCore::RenderSVGResourceRadialGradient::createGradient):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceLinearGradient.cpp:
(WebCore::LegacyRenderSVGResourceLinearGradient::buildGradient const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceRadialGradient.cpp:
(WebCore::LegacyRenderSVGResourceRadialGradient::buildGradient const):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.cpp:
(WebKit::RemoteGraphicsContext::setFillCachedGradient):
(WebKit::RemoteGraphicsContext::setStrokeCachedGradient):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheGradient):
(WebKit::RemoteRenderingBackend::releaseGradient):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheGradient):
(WebKit::RemoteResourceCache::releaseGradient):
(WebKit::RemoteResourceCache::cachedGradient const):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::recordResourceUse):
(WebKit::RemoteGraphicsContextProxy::appendStateChangeItemIfNecessary):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheGradient):
(WebKit::RemoteRenderingBackendProxy::releaseGradient):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordGradientUse):
(WebKit::RemoteResourceCacheProxy::willDestroyGradient):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: https://commits.webkit.org/300094@main
